File: src/main/java/org/stt/gui/jfx/SettingsController.java
package org.stt.gui.jfx;

import javafx.application.Platform;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.property.StringProperty;
import javafx.beans.value.ObservableValue;
import javafx.scene.Node;
import javafx.scene.control.PasswordField;
import javafx.scene.control.TextField;
import javafx.scene.control.TextInputControl;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.Pane;
import javafx.util.Callback;
import org.controlsfx.control.PropertySheet;
import org.controlsfx.property.BeanPropertyUtils;
import org.controlsfx.property.editor.AbstractPropertyEditor;
import org.controlsfx.property.editor.PropertyEditor;
import org.stt.config.*;

import javax.inject.Inject;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class SettingsController {
    private final ConfigRoot configRoot;
    private final ActivitiesConfig activitiesConfig;
    private final BackupConfig backupConfig;
    private final CliConfig cliConfig;
    private final CommonPrefixGrouperConfig commonPrefixGrouperConfig;
    private final ReportConfig reportConfig;
    private final WorktimeConfig worktimeConfig;
    private final JiraConfig jiraConfig;
    private Pane pane;

    @Inject
    public SettingsController(ConfigRoot configRoot, // NOSONAR
                              ActivitiesConfig activitiesConfig,
                              BackupConfig backupConfig,
                              CliConfig cliConfig,
                              CommonPrefixGrouperConfig commonPrefixGrouperConfig,
                              ReportConfig reportConfig,
                              WorktimeConfig worktimeConfig,
                              JiraConfig jiraConfig) {
        this.configRoot = configRoot;
        this.activitiesConfig = activitiesConfig;
        this.backupConfig = backupConfig;
        this.cliConfig = cliConfig;
        this.commonPrefixGrouperConfig = commonPrefixGrouperConfig;
        this.reportConfig = reportConfig;
        this.worktimeConfig = worktimeConfig;
        this.jiraConfig = jiraConfig;
    }

    private PropertyEditor<?> createPasswordSettingsEditor(PropertySheet.Item property) {
        AbstractPropertyEditor<PasswordSetting, PasswordField> propertyEditor = new AbstractPropertyEditor<PasswordSetting, PasswordField>(property, new PasswordField()) {
            @Override
            protected ObservableValue<PasswordSetting> getObservableValue() {
                StringProperty textProperty = getEditor().textProperty();
                ObjectProperty<PasswordSetting> passwordSettingProperty = new SimpleObjectProperty<>();
                applyValue(passwordSettingProperty, textProperty.getValue());
                textProperty.addListener((observable, oldValue, newValue) -> applyValue(passwordSettingProperty, newValue));
                return passwordSettingProperty;
            }

            private void applyValue(ObjectProperty<PasswordSetting> passwordSettingProperty, String newValue) {
                passwordSettingProperty.setValue(newValue == null ? null : PasswordSetting.fromPassword(newValue.getBytes(StandardCharsets.UTF_8)));
            }

            @Override
            public void setValue(PasswordSetting value) {
                getEditor().setText(value == null ? null : new String(value.getPassword(), StandardCharsets.UTF_8));
            }
        };
        enableAutoSelectAll(propertyEditor.getEditor());
        return propertyEditor;
    }

    private static AbstractPropertyEditor<PathSetting, TextField> createPathSettingsEditor(PropertySheet.Item property) {
        AbstractPropertyEditor<PathSetting, TextField> propertyEditor = new AbstractPropertyEditor<PathSetting, TextField>(property, new TextField()) {
            @Override
            protected ObservableValue<PathSetting> getObservableValue() {
                StringProperty textProperty = getEditor().textProperty();
                ObjectProperty<PathSetting> pathSettingProperty = new SimpleObjectProperty<>();
                applyValue(pathSettingProperty, textProperty.getValue());
                textProperty.addListener((observable, oldValue, newValue) -> applyValue(pathSettingProperty, newValue));
                return pathSettingProperty;
            }

            private void applyValue(ObjectProperty<PathSetting> pathSettingProperty, String newValue) {
                pathSettingProperty.setValue(new PathSetting(newValue));
            }

            @Override
            public void setValue(PathSetting value) {
                getEditor().setText(value.path());
            }
        };
        enableAutoSelectAll(propertyEditor.getEditor());
        return propertyEditor;
    }

    private static void enableAutoSelectAll(final TextInputControl control) {
        control.focusedProperty().addListener((ObservableValue<? extends Boolean> o, Boolean oldValue, Boolean newValue) -> {
            if (newValue) {
                Platform.runLater(control::selectAll);
            }
        });
    }

    public Node getPanel() {
        initialize();
        return pane;
    }

    private void initialize() {
        PropertySheet propertySheet = new PropertySheet();
        Callback<PropertySheet.Item, PropertyEditor<?>> defaultEditorFactory = propertySheet.getPropertyEditorFactory();
        propertySheet.setPropertyEditorFactory(item -> {
            Class<?> type = item.getType();
            if (PathSetting.class.isAssignableFrom(type)) {
                return createPathSettingsEditor(item);
            }
            if (PasswordSetting.class.isAssignableFrom(type)) {
                return createPasswordSettingsEditor(item);
            }
            return defaultEditorFactory.call(item);
        });
        Stream.of(configRoot, activitiesConfig, backupConfig, cliConfig, commonPrefixGrouperConfig,
                reportConfig, worktimeConfig, jiraConfig)
                .forEach(o -> propertySheet.getItems().addAll(BeanPropertyUtils.getProperties(o,
                        propertyDescriptor -> {
                            Class<?> propertyType = propertyDescriptor.getPropertyType();
                            return !(ConfigurationContainer.class.isAssignableFrom(propertyType));
                        })));
        pane = new BorderPane(propertySheet);
    }
}

File: src/main/kotlin/org/stt/gui/jfx/SettingsController.kt
package org.stt.gui.jfx

import javafx.application.Platform
import javafx.beans.property.ObjectProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.beans.value.ObservableValue
import javafx.scene.Node
import javafx.scene.control.PasswordField
import javafx.scene.control.TextField
import javafx.scene.control.TextInputControl
import javafx.scene.layout.BorderPane
import javafx.scene.layout.Pane
import org.controlsfx.control.PropertySheet
import org.controlsfx.property.BeanPropertyUtils
import org.controlsfx.property.editor.AbstractPropertyEditor
import org.controlsfx.property.editor.PropertyEditor
import org.stt.config.*
import java.nio.charset.StandardCharsets
import java.util.stream.Stream
import javax.inject.Inject

class SettingsController @Inject
constructor(private val configRoot: ConfigRoot, // NOSONAR
            private val activitiesConfig: ActivitiesConfig,
            private val backupConfig: BackupConfig,
            private val cliConfig: CliConfig,
            private val commonPrefixGrouperConfig: CommonPrefixGrouperConfig,
            private val reportConfig: ReportConfig,
            private val worktimeConfig: WorktimeConfig,
            private val jiraConfig: JiraConfig) {
    private var pane: Pane? = null

    val panel: Node?
        get() {
            initialize()
            return pane
        }

    private fun createPasswordSettingsEditor(property: PropertySheet.Item): PropertyEditor<*> {
        val propertyEditor = object : AbstractPropertyEditor<PasswordSetting, PasswordField>(property, PasswordField()) {
            override fun getObservableValue(): ObservableValue<PasswordSetting> {
                val textProperty = editor.textProperty()
                val passwordSettingProperty = SimpleObjectProperty<PasswordSetting>()
                applyValue(passwordSettingProperty, textProperty.value)
                textProperty.addListener { _, _, newValue -> applyValue(passwordSettingProperty, newValue) }
                return passwordSettingProperty
            }

            private fun applyValue(passwordSettingProperty: ObjectProperty<PasswordSetting>, newValue: String?) {
                passwordSettingProperty.value = if (newValue == null) null else PasswordSetting.fromPassword(newValue.toByteArray(StandardCharsets.UTF_8))
            }

            override fun setValue(value: PasswordSetting?) {
                editor.text = if (value == null) null else String(value.password, StandardCharsets.UTF_8)
            }
        }
        enableAutoSelectAll(propertyEditor.editor)
        return propertyEditor
    }

    private fun createPathSettingsEditor(property: PropertySheet.Item): AbstractPropertyEditor<PathSetting, TextField> {
        val propertyEditor = object : AbstractPropertyEditor<PathSetting, TextField>(property, TextField()) {
            override fun getObservableValue(): ObservableValue<PathSetting> {
                val textProperty = editor.textProperty()
                val pathSettingProperty = SimpleObjectProperty<PathSetting>()
                applyValue(pathSettingProperty, textProperty.value)
                textProperty.addListener { _, _, newValue -> applyValue(pathSettingProperty, newValue) }
                return pathSettingProperty
            }

            private fun applyValue(pathSettingProperty: ObjectProperty<PathSetting>, newValue: String) {
                pathSettingProperty.value = PathSetting(newValue)
            }

            override fun setValue(value: PathSetting) {
                editor.text = value.path()
            }
        }
        enableAutoSelectAll(propertyEditor.editor)
        return propertyEditor
    }

    private fun enableAutoSelectAll(control: TextInputControl) {
        control.focusedProperty().addListener { _: ObservableValue<out Boolean>, _: Boolean, newValue: Boolean ->
            if (newValue) {
                Platform.runLater { control.selectAll() }
            }
        }
    }

    private fun initialize() {
        val propertySheet = PropertySheet()
        val defaultEditorFactory = propertySheet.propertyEditorFactory
        propertySheet.setPropertyEditorFactory { item ->
            val type = item.type
            if (PathSetting::class.java.isAssignableFrom(type)) {
                return@setPropertyEditorFactory createPathSettingsEditor(item)
            }
            if (PasswordSetting::class.java.isAssignableFrom(type)) {
                return@setPropertyEditorFactory createPasswordSettingsEditor(item)
            }
            defaultEditorFactory.call(item)
        }
        Stream.of(configRoot, activitiesConfig, backupConfig, cliConfig, commonPrefixGrouperConfig,
                reportConfig, worktimeConfig, jiraConfig)
                .forEach { o ->
                    propertySheet.items.addAll(BeanPropertyUtils.getProperties(o
                    ) { propertyDescriptor ->
                        val propertyType = propertyDescriptor.propertyType
                        !ConfigurationContainer::class.java.isAssignableFrom(propertyType)
                    })
                }
        pane = BorderPane(propertySheet)
    }
}
